axlearn/common/checkpointer_orbax_emergency_replicator.py

```diff
@@ -0,0 +1,346 @@
+                `save_policy`. Note that data iterator will be saved with either `save_policy` or
+                `local_save_policy` indicate we should save.
+            async_timeout_secs: Timeout for async barrier in seconds when saving tensors.
+            replica_axis_index: The index of the "data" axis.
+        """
+
+        keep_last_n: int = 1
+        keep_every_n_steps: Optional[int] = None
+        local_keep_last_n: int = 1
+        local_save_policy: InstantiableConfig[CheckpointPolicy] = config_for_function(
+            every_n_steps_policy
+        ).set(n=10)
+        local_dir: str = "/checkpoint"
+        trainer_dir: Required[str] = REQUIRED
+        async_timeout_secs: int = 3600
+        replica_axis_index: Required[int] = REQUIRED
+
+    def __init__(self, cfg: Config, *, parent: Optional[Module]):
+        super().__init__(cfg, parent=parent)
+        cfg: OrbaxEmergencyReplicatorCheckpointer.Config = self.config
+        self._name_format = ocp.step.standard_name_format(
+            step_prefix=None,
+            step_format_fixed_length=None,
+        )
+        self._local_dir = cfg.local_dir
+        # Orbax replicator ckpt requires this function to be called prior to checkpointer
+        # operations. This function also serves as a barrier.
+        ocp.multihost.initialize_runtime_to_distributed_ids()
+        ocp.multihost.initialize_distributed_to_device_ids()
+
+        num_slices = int(os.environ["MEGASCALE_NUM_SLICES"])
+
+        replicator_file = "replicator.yaml"
+        temp_file = replicator_file + ".tmp"
+        replicator_file_path = epath.Path(self._local_dir) / replicator_file
+        if not _wait_for_file_to_disappear(replicator_file_path):
+            logging.info("Existing replicator.yaml did not disappear in time.")
+        else:
+            logging.info("replicator.yaml no longer exists, creating new replicator.yaml.")
+        temp_file = epath.Path(self._local_dir) / temp_file
+        num_nodes = jax.process_count()
+        nodes_per_slice = num_nodes // num_slices
+
+        node_rank = global_state.process_id
+        my_process_index = jax.process_index()
+        proc_index_to_node_rank = ocp.multihost.runtime_to_distributed_ids()
+
+        my_in_pipeline_index = my_process_index % nodes_per_slice
+        peer_ranks = []
+        for i in range(num_slices):
+            peer_process_index = i * nodes_per_slice + my_in_pipeline_index
+            if peer_process_index != my_process_index:
+                peer_process_rank = proc_index_to_node_rank[peer_process_index]
+                peer_ranks.append(peer_process_rank)
+
+        logging.info("Peers for NodeRank %s: %s", node_rank, peer_ranks)
+
+        run_name = os.environ.get("HOSTNAME").split("job")[0].rstrip("-")
+
+        replicator_yaml = f"""job-name: {run_name}
+      framework: orbax
+      assume-data-parallelism: 2
```

Comment:
@DmitryKakurin how do we set this correctly in the following scenarios?

1. ici_dp=8, dcn_dp=4, should it be 32?
2. ici_dp=1 dcn_dp=4

seems in maxtext it's hardcoded as num_slices which isn't always the case in axlearn: https://github.com/AI-Hypercomputer/maxtext/blob/946f2a60c2c05f8fe8540fefb7e44b3838f87eb3/MaxText/max_utils.py#L306

Reply:
This should come as a config param in `Config` class.

Reply:
Still unclear what to use when ici_dp is bigger than 1